### DISCUSSION TITLE
Allow stepper direction config ($3) for ganged motors.

### DISCUSSION THREAD

**Discussion by phil-barrett** [grblHAL/core#90] [Ideas]
When ganging two motors together it is often necessary to reverse one of the motor's direction.

For stepper motors, this is fairly easy at the wiring level - just reverse one of the motor wire pairs.  For servos, it requires programming the servo.  I think it would be a good addition to allow $3 to also change the step direction for the 2nd motor in the ganged pair.  This would allow builders to not have to change their hardware setup once installed. Here's an ioSender mockup to show the idea.

I think the defaults should be to invert the ganged drive step direction. Y not inverted, B inverted for the example shown  Also, if inverting main axis (Y in this case), the ganged drive should have inversion applied to it as well. Thus, if Y was originally not inverted but B was and subsequently the user inverts Y, B would be inverted again (ie, show as not inverted).

Note that the image below implies changes to grblHAL and ioSender, I would be happy with just treating it like an extra axis (i.e., no sender changes, just show the extra axis). 

![image](https://user-images.githubusercontent.com/16570958/142737858-41f131d4-c569-4c9f-8759-dc60bccacc4c.png)

**Comment by terjeio** [comment]
IMO a new setting is called for that inverts the $3 direction setting for the ganged motor instead of setting it explicitly. This would be far simpler to implement as having it part of the $3 setting will trigger a lot of code changes that I am quite sure I am not going to even try to do.
Also, a new setting will not require any changes in the sender and neither in the settings descriptors used by the sender for creating the setting UI elements - a big plus.

BTW I now label motors > 3 as M3, M4 and M5 instead of A, B, C in the map files. Perhaps this should be adapted for the board markings as well?
For me it is confusing to associate the B axis with the ganged Y motor. In your example above A = M3 and the ganged Y = M4.
A new $-command to show the motor mappings would solve this issue?

[edit: changed B = M3 to A = M3 above]

**Reply by phil-barrett** [reply]
That works for me. Hopefully it handles the case where there are two sets of ganged motors.  A dual core-XY hot wire foam cutter is an example of that.

By the way, I think M3=Z, M4=A and M5 is associated with Y (M2+M5?).

And, I totally agree with the M3-5 idea. It really helps with the ambiguity of Axis vs Motor. In the 3 Axis ganged situation there are 4 motors and 3 axes.  Using M? allows a clear distinction.

**Reply by terjeio** [reply]
[New version](https://github.com/grblHAL/core/blob/master/changelog.md) committed:

> Hopefully it handles the case where there are two sets of ganged motors. 

Up to three is supported (X, Y and Z) - here is with two enabled:
![image](https://user-images.githubusercontent.com/20260062/142770133-36442826-21ad-45ce-93d8-88a606cb47cf.png)

I have _not_ tested this in a machine, so please test carefully and report back.

The description should be improved but I got lazy in the end as I wanted to spend the day in the workshop but ended up implementing this instead - mainly because I was ready to commit an update with settings changes before this came up. So perfect timing by you.

**Reply by phil-barrett** [reply]
Wow, that was fast!  Thank you.  I hear you about getting into the workshop.  Way too many things crying for attention.  I will give it a test but probably not for a couple of days.

Thank you for you dedication to grblHAL.
